File: src/internal/computer/operand.py
# -*- coding: utf-8 -*-

from enum import Enum
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .state import ComputerState


class OperandType(Enum):
    Reg = 0
    Mem = 1
    Imm = 2
    Str = 3


class Operand:
    reg: str = ""
    offset: int = 0
    immVal: int = 0
    text: str = ""

    def __init__(self, tp: OperandType, reg: str = "", offset: int = 0, immVal: int = 0, text: str = ""):
        self.tp = tp
        self.reg = reg
        self.offset = offset
        self.immVal = immVal
        self.text = text

    def mem_addr(self, state: 'ComputerState'):
        if self.tp != OperandType.Mem:
            raise TypeError("Operand type must be Mem")
        v = state.get_reg_value(self.reg)
        return v + self.offset

    def value(self, state: 'ComputerState') -> int | str:
        if self.tp == OperandType.Reg:
            return state.get_reg_value(self.reg)
        elif self.tp == OperandType.Mem:
            return state.get_reg_mem_value(self.reg, self.offset)
        elif self.tp == OperandType.Imm:
            return self.immVal
        elif self.tp == OperandType.Str:
            return self.text
        else:
            raise TypeError(f"invalid Operand type:{self.tp}")

    def __repr__(self):
        if self.tp == OperandType.Reg:
            return self.reg
        elif self.tp == OperandType.Mem:
            if self.offset == 0:
                return f"[{self.reg}]"
            else:
                return f"[{self.reg}{self.offset:+d}]"
        elif self.tp == OperandType.Imm:
            return str(self.immVal)
        elif self.tp == OperandType.Str:
            return f'"{self.text}"'
        else:
            raise TypeError(f"invalid Operand type:{self.tp}")

File: src/web_frontend/src/libs/Operand.ts
export enum OperandType {
    Reg = "Reg",
    Mem = "Mem",
    Imm = "Imm",
    Str = "Str"
}

export class Operand {
    constructor(
        public tp: OperandType,
        public reg: string = "",
        public offset: number = 0,
        public immVal: number = 0,
        public text: string[] = [],
    ) {}

    /**
     * Get the memory address for a memory operand
     */
    memAddr(registers: { [key: string]: number }): number {
        if (this.tp !== OperandType.Mem) {
            throw new Error("Operand type must be Mem");
        }
        const regValue = registers[this.reg];
        return regValue + this.offset;
    }

    /**
     * Get the value of this operand
     */
    value(registers: { [key: string]: number }, memory: number[]): number | string {
        switch (this.tp) {
            case OperandType.Reg:
                return registers[this.reg];
            case OperandType.Mem:
                const addr = this.memAddr(registers);
                if (addr < 0 || addr >= memory.length) {
                    throw new Error(`内存越界: ${addr}`);
                }
                return memory[addr];
            case OperandType.Imm:
                return this.immVal;
            case OperandType.Str:
                return this.text[1];
            default:
                throw new Error(`Invalid Operand type: ${this.tp}`);
        }
    }

    toString(): string {
        switch (this.tp) {
            case OperandType.Reg:
                return this.reg;
            case OperandType.Mem:
                if (this.offset === 0) {
                    return `[${this.reg}]`;
                } else {
                    return `[${this.reg}${this.offset >= 0 ? '+' : ''}${this.offset}]`;
                }
            case OperandType.Imm:
                return String(this.immVal);
            case OperandType.Str:
                return this.text[0];
            default:
                throw new Error(`Invalid Operand type: ${this.tp}`);
        }
    }
}
